### PR TITLE
Fix Ability Activation Order for type changes

### PR DESCRIPTION
**Pull request by ascriptmaster**
~~Hidden Power is unaffected by Normalize.~~

~~Also a researcher should check if it's the case with Judgment and Techno Blast.~~

Fixed ability interactions with Hidden Power, Judgment, and other moves
that change their types.

Normalize > move's type change > -ate Ability > Ion Deluge/Electrify

For Gen 4, the order is: move's type change > Normalize

### CODE REVIEW

**Comment by TheImmortal** [comment]
Uhm, I think you misread cause this is what Veekun says:
>Moves of variable type, such as Hidden Power, are affected.

**Comment by ascriptmaster** [comment]
> affected
Well then, I guess I have to inform the person who made the bug report

Bulbapedia says otherwise but Bulbapedia is dumb. Is Veekun the most accurate?

**Comment by TheImmortal** [comment]
More accurate than Bulbapedia. Less accurate than @Marty-D.

**Comment by Marty-D** [comment]
They're both half right. Veekun is right for Gen 4 and Bulbapedia is right for Gen 5 and 6.

Gen 6 goes like this:

listed move type -> moves that call other moves use the new move instead -> Normalize changes the move to Normal -> moves now change type (Hidden Power/Judgment/Natural Gift/Techno Blast/Weather Ball) -> if the move is unchanged and still Normal (i.e. not Natural Gift with Chilan Berry), Aerilate/Pixilate/Refrigerate change it to Flying/Fairy/Ice -> if the move is Normal, Ion Deluge changes to Electric -> Electrify changes the move to Electric -> Protean activates -> Gems activate if the Gem matches the move type

Gen 5 is the same (but obviously the effects that didn't exist then don't apply).

For Gen 4, it's just: listed move type -> moves that call other moves use the new move instead -> moves now change type (Hidden Power/Judgment/Natural Gift/Weather Ball) -> Normalize changes the move to Normal

Struggle stays typeless in all cases.

**Comment by TheImmortal** [comment]
Thanks Marty! @ascriptmaster I guess you need to change the gen 4 mod as well.

File: data/abilities.js [hunk 1]
```diff
@@ -1753,7 +1753,7 @@ exports.BattleAbilities = {
 		desc: "Makes all of this Pokemon's attacks Normal-typed.",
 		shortDesc: "This Pokemon's moves all become Normal-typed.",
 		onModifyMove: function (move) {
-			if (move.id !== 'struggle') {
+			if (move.id !== 'struggle' && move.id.substr(0, 11) !== 'hiddenpower') {
```
Comment (Slayer95) [inline data/abilities.js]: You may just check <code>move.id === ...</code>

**Comment by ascriptmaster** [comment]
Implementation probably goes something like this?

(I didn't know what to do with Fling, which originally had -1 on `ModifyMovePriority`, so I pushed its priority backwards so that these type-change abilities would go first.)

**Comment by ascriptmaster** [comment]
@Marty-D is an amazing person who reminded me that Natural Gift is designed not to interact with -ate abilities, so it gets coded in as an exception.

**Pull request merged**